### ISSUE TITLE
date parsing is not aware of non-english month names

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#1874]
Reported by pablo1107 in #plaintextaccounting: a `date-format` in CSV rules did not perform as expected:
```
~/ledger master*
❯ export LC_TIME="es_AR.UTF-8"

~/ledger master*
❯ hledger print --rules-file vacaciones-tc.rules -f vacaciones-tc.csv
hledger: error: could not parse "30 abr 2022" as a date using date format "%d %h %Y"
record values: "30 abr 2022","MERCADOPAGO*QUILMESROCK          02/06 ","ARS  +383,33"
the date rule is:   %1
the date-format is: %d %h %Y
you may need to change your date rule, change your date-format rule, or change your skip rule
for m/d/y or d/m/y dates, use date-format %-m/%-d/%Y or date-format %-d/%-m/%Y
```

In fact hledger has no awareness of the system time locale / $LC_TIME; it hard codes the en_US time locale (as do essentially all Haskell programs, I would guess). 

This makes it difficult to parse CSV containing non-english month/day names. Possibly it also manifests in other ways, though I don't know of any.

https://hackage.haskell.org/package/env-locale seems to be the way to get the system time locale. Here's a test I didn't get to work for me:
```haskell
#!/usr/bin/env stack
-- stack runghc --verbosity info --package time --package env-locale

import Data.Time
import Data.Time.Format
import System.Locale.Current

main = do
  ctl <- currentLocale 
  d <- parseTimeM False ctl "%b" "abr" :: IO Day
  print d
```

This is a slight can of worms though, similar to the text encoding discussion at #1834 etc. Mainly, if this capability is desirable, and it seems so, what is the best default providing predictability, convenience for reading your local data, and convenience for reading foreign-language data. Probably we can follow however we decide to handle text encoding.

**Comment by sten0** [comment]
> Mainly, if this capability is desirable, and it seems so, what is the best default providing predictability, convenience for reading your local data, and convenience for reading foreign-language data. Probably we can follow however we decide to handle text encoding.

I think the expectation of most programs is that they follow LANG, LANGUAGE, with possibly optional LC_TIME exception.  It seems like it might be optional whether or not this needs to be made explicit in the primary journal file.

That said, the language of banking might be a lingua franca.  It makes sense that foo.csv.rules would define the input language for foo.csv.

The third concern is when the input language is different than the output language, whether from environmental variables, or as-defined in the journal file that is importing a foreign foo.csv.  It seems to me that the principle of least surprise here is that the output to the journal file is localised.  When foreign csv files are archived, they should be archived as-is and not localised.

**Comment by simonmichael** [comment]
I agree with where you're going @sten0. Input files can be using any conventions, so there needs to be a way to specify those conventions so the files can be read. Output can reasonably, in some cases, be based on system locale (& ideally easily overrideable).

A next question is what exactly in hledger should be time-locale-sensitive, or general-locale-sensitive ?

This issue mentions month names parsed by csv rules specifically; let's keep it focussed on csv reading. The date-format rule provides flexibility for parsing date-times, but not enough; allowing it to parse non-english short and long month names would be a step forward. The locale to use should be configurable in the rules file.

What else might need to be locale-sensitive in date-format's pattern, or when reading CSV files ?

**Comment by sten0** [comment]
Sorry for this wordy reply.  I could maybe make it, at best, 10% shorter
by taking 30min additional time.  If you want to skip to one immediately
actionable item/bug, please skip to the end.

Simon Michael ***@***.***> writes:

> simonmichael left a comment (simonmichael/hledger#1874)
>
> I agree with where you're going @sten0. Input files can be using any conventions, so there needs to be a way to specify those conventions so the files can be read. Output can reasonably, in some cases, be based on system locale (& ideally easily overrideable).

For output, I'm not sure if environment variables should take precedence
over what is configured in file.  For example, for bug reporting it
often requested to set LANG and LC_ALL to C.UTF-8 for a project that
uses English, so the developers can quickly understand the debugging
output.  I have to defer to you for what makes sense to implement.
Ideally it would be nice if hledger could know if it was outputing to
stdout (and should thus use the envvar locale) or to a file.  In the
case of a journal file, it would be unexpected for some txns to be in
English because the user forgot to disable the envvar used for
bugreporting.  It seems like the way to guard against this is to parse
the journal to find a target language.

> A next question is what exactly in hledger should be time-locale-sensitive, or general-locale-sensitive ?
>

A case I can think of is that the user's primary bank is something like
a Catalan Credit Union (presumably locale=Catalan), but the Madrid stock
exchange uses Castillian Spanish.  Not to mention NYSE international
trades, of course!  Must the user have three journals?  Alternatively,
is is unreasonable to have a journal that uses a single date format that
contains txn descriptions in different languages?

Tags are user-defined and thus unlocalised, because this is an unbounded
problem.

Ah, considering different stock exchange timezones means that localising
transation time accross the international dateline creates a problem.  I
think the only.  Maybe this will solve it (if it hasn't already been
solved):

1. Document that when multiple timezones are used, each csv.rules
should specify the timezone the txn occcured in.
2. hledger needs to be able to track GMT, plus offsets.  Convenience
aliases are a plus, of course.
3. When a csv is imported, tag the timezone in the journal, and record
the ISO-converted date that will match the bank (or exchange) statement.
4. Maybe provide a mechanism for hledger to normalise any txn or list of
txns to user-time-zone?
5. Corner case: What happens when the user manually enters txns on both
sides of the international date line within the same 12 period?  In
other words, do all txns need to be datestamped?  Maybe allow a opt-in
"home" timezone, and any tz outside of that gets tz tag?

> This issue mentions month names parsed by csv rules specifically; let's keep it focussed on csv reading. The date-format rule provides flexibility for parsing date-times, but not enough; allowing it to parse non-english short and long month names would be a step forward. The locale to use should be configurable in the rules file.

Sounds good to me!

> What else might need to be locale-sensitive in date-format's pattern, or when reading CSV files ?

date-format needs to support punctuation, and currently this fails with:

```
hledger: Error: could not parse "June 30, 2026" as a date using date format "%B, %-d %Y"
record: June 30 2026,
the date rule is:   %1
the date-format is: %B, %-d %Y
```
